src/physics/material/MaterialParams.cc

```diff
@@ -0,0 +1,243 @@
+    result.materials = make_span(host_materials_);
+    ENSURE(result);
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+
+/*!
+ * Access material properties on the device.
+ */
+MaterialParamsPointers MaterialParams::device_pointers() const
+{
+    MaterialParamsPointers result;
+    result.elements  = device_elements_.device_pointers();
+    result.materials = device_materials_.device_pointers();
+    ENSURE(result);
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+// IMPLEMENTATION
+//---------------------------------------------------------------------------//
+/*!
+ * Convert an element input to an element definition.
+ *
+ * This adds computed quantities in addition to the input values. The result
+ * is pushed back onto the
+ */
+void MaterialParams::append_element_def(const ElementInput& inp)
+{
+    REQUIRE(inp.atomic_number > 0);
+    REQUIRE(inp.atomic_mass > zero_quantity());
+
+    ElementDef result;
+
+    // Copy basic properties
+    result.atomic_number = inp.atomic_number;
+    result.atomic_mass   = inp.atomic_mass;
+
+    // Calculate various factors of the atomic number
+    const real_type z_real      = result.atomic_number;
+    result.cbrt_z               = std::cbrt(z_real);
+    result.cbrt_zzp             = std::cbrt(z_real * (z_real + 1));
+    result.log_z                = std::log(z_real);
+    result.mass_radiation_coeff = detail::calc_mass_rad_coeff(result);
+
+    elnames_.push_back(inp.name);
+
+    // Add to host vector
+    host_elements_.push_back(result);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Process and add element components to the internal list.
+ *
+ * \todo It's the caller's responsibility to ensure that element IDs
+ * aren't duplicated.
+ */
+span<MatElementComponent>
+MaterialParams::extend_elcomponents(const MaterialInput& inp)
+{
+    REQUIRE(host_elcomponents_.size() + inp.elements.size()
+            <= host_elements_.capacity());
+
+    // Allocate material components
+    auto start_size = host_elcomponents_.size();
+    host_elcomponents_.resize(start_size + inp.elements.size());
+    span<MatElementComponent> result{host_elcomponents_.data() + start_size,
+                                     inp.elements.size()};
+
+    // Copy values from input and check number fractions
+    real_type norm = 0.0;
+    for (auto i : range(inp.elements.size()))
+    {
+        REQUIRE(inp.elements[i].first < host_elements_.size());
+        REQUIRE(inp.elements[i].second >= 0);
+        result[i].element  = inp.elements[i].first;
+        result[i].fraction = inp.elements[i].second;
+        norm += inp.elements[i].second;
+    }
+    CHECK(SoftEqual<>()(norm, 1.0) || inp.elements.empty());
+    norm = 1 / norm;
+
+    // Normalize
+    for (MatElementComponent& comp : result)
+    {
+        comp.fraction *= norm;
+    }
```

Comment:
Shouldn't this responsibility be left to whoever defines the geometry?

Reply:
Yeah, i was of two minds about how much processing vs checking vs assuming here. I figured Geant might not be as rigorous as we’d like so it would be ok to do a cheap renormalization

Reply:
Since this happens only during initialization, I agree that it can't do any harm. Although it seems very unlikely, I wonder if small rounding differences may alter our final distributions when comparing/validating with Geant...

Reply:
There are probably going to be plenty of differences between us and geant4 that are much greater than the potential roundoff... here I'm more concerned about ensuring our stuff is internally consistent here.